compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/gen_c_transformer/expr_transformer.rs

```diff
@@ -0,0 +1,277 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use std::ops::{BitAnd, Shl, Shr};
+
+use super::super::super::super::MachineModel;
+use super::super::super::{
+    BinaryOperand, CIntType, Expr, Location, Parameter, Stmt, Symbol, SymbolTable, SymbolValues,
+    Type,
+};
+use super::super::Transformer;
+use super::common::type_to_string;
+use num::bigint::BigInt;
+use rustc_data_structures::fx::FxHashMap;
+
+/// Create an expr from an int constant using only values <= u64::MAX.
+fn bignum_to_expr(num: &BigInt, typ: &Type) -> Expr {
+    // CInteger types should already be valid in C
+    if matches!(typ, Type::CInteger(_)) {
+        return Expr::int_constant(num.clone(), typ.clone());
+    }
+
+    // Only need to handle type wider than 64 bits
+    if let Some(width) = typ.width() {
+        if width <= 64 {
+            return Expr::int_constant(num.clone(), typ.clone());
+        }
+    }
+
+    // Only types that should be left are i128 and u128
+    assert_eq!(typ.width(), Some(128), "Unexpected int width: {:?}", typ.width());
+
+    let left_mask = BigInt::from(u64::MAX).shl(64);
+    let right_mask = BigInt::from(u64::MAX);
+
+    let unsigned_typ = match typ {
+        Type::Unsignedbv { .. } => typ.clone(),
+        Type::Signedbv { width } => Type::unsigned_int(*width),
+        _ => panic!("Unexpected type in bignum_to_expr: {:?}", typ),
+    };
+    let left_half: BigInt = num.bitand(left_mask);
+    let right_half = num.bitand(right_mask);
+
+    Expr::int_constant(left_half.shr(64), unsigned_typ.clone())
+        .shl(Expr::int_constant(64, Type::c_int()))
+        .bitor(Expr::int_constant(right_half, unsigned_typ))
+        .cast_to(typ.clone())
+}
+
+/// Struct for handling the expression replacement transformations for --gen-c-runnable.
+pub struct ExprTransformer {
+    new_symbol_table: SymbolTable,
+    empty_statics: FxHashMap<String, Expr>,
+}
+
+impl ExprTransformer {
+    /// Replace expressions which lead to invalid C with alternatives.
+    pub fn transform(original_symbol_table: &SymbolTable) -> SymbolTable {
+        let new_symbol_table = SymbolTable::new(original_symbol_table.machine_model().clone());
+        ExprTransformer { new_symbol_table, empty_statics: FxHashMap::default() }
+            .transform_symbol_table(original_symbol_table)
+    }
+
+    /// Extract `empty_statics` map for final processing.
+    pub fn empty_statics_owned(&mut self) -> FxHashMap<String, Expr> {
+        std::mem::replace(&mut self.empty_statics, FxHashMap::default())
+    }
+
+    /// Add identifier to a transformed parameter if it's missing.
+    /// Necessary when function wasn't originally a definition.
```

Comment:
e.g. extern functions. Explain why you need to give them function bodies.

Reply:
Done.